src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php

```diff
@@ -490,6 +532,42 @@ private function createAuthenticationListeners(ContainerBuilder $container, stri
         return [$listeners, $defaultEntryPoint];
     }
 
+    private function getUserProvider(ContainerBuilder $container, string $id, array $firewall, string $factoryKey, ?string $defaultProvider, array $providerIds, ?string $contextListenerId): ?string
+    {
+        if (isset($firewall[$factoryKey]['provider'])) {
+            if (!isset($providerIds[$normalizedName = str_replace('-', '_', $firewall[$factoryKey]['provider'])])) {
+                throw new InvalidConfigurationException(sprintf('Invalid firewall "%s": user provider "%s" not found.', $id, $firewall[$factoryKey]['provider']));
+            }
+
+            return $providerIds[$normalizedName];
+        }
+
+        if ('remember_me' === $factoryKey || 'anonymous' === $factoryKey) {
+            if ('remember_me' === $factoryKey && $contextListenerId) {
+                $container->getDefinition($contextListenerId)->addTag('security.remember_me_aware', ['id' => $id, 'provider' => 'none']);
+            }
```

Comment:
In the NEW system, it would be great to get rid of this. There is already a `DeauthenticatedEvent` that ContextListener dispatches... so in theory, the `RemeberMeFactory::createAuthenticator()` could register a listener on this (the listener would need to be aware of its firewall id... so that it fired only when that event happened in ITs firewall).

But, DeauthenticatedEvent is not *exactly* the same as when `$this->rememberMeServices->loginFail()` - there is a subtle difference between the event and when `loginFail()` is called (`loginFail` is called in more situations). So to get this to work, we may need a 2nd, new event. And maybe it's even some new `LogoutEvent` ... because that's really what's happening inside `ContextListener` in this situation.

Reply:
Continued in #40145